# __Hands-on with TensorFlow: Part A__

Let's see how TensorFlow works and build a deep neural network model using the MNIST dataset.

## Steps to be followed:
1. Import TensorFlow
2. Load the MNIST dataset
3. Create the model
4. Get predictions from the model
5. Apply softmax activation to the predictions
6. Define the loss function
7. Compile the model
8. Train the model
9. Evaluate the model
10. Create a probability model

## Step 1: Import TensorFlow

- Import TensorFlow and check its version.


In [ ]:
import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'

import tensorflow as tf
print("TensorFlow version:", tf.__version__)

TensorFlow version: 2.15.0


## Step 2: Load the MNIST dataset

- Load the MNIST dataset and normalize the input data by dividing the train and test sets by 255. The pixel values of the images are divided by 255.0 to normalize them to the range [0, 1]



The MNIST dataset is a large collection of handwritten digits commonly used for training and testing in the field of machine learning. It is one of the most well-known datasets for evaluating the performance of image processing and pattern recognition algorithms.

In [ ]:
mnist = tf.keras.datasets.mnist

(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

11490434/11490434 [==============================] - 0s 0us/step


__Observation:__
- The dataset has been downloaded.

## Step 3: Create the model

- Create a Sequential model with flatten, dense, and dropout layers.


In [ ]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10)
])

## Step 4: Get predictions from the model

- Get the predictions from the model using the train data for one column, as the optimizer is not yet applied.


In [ ]:
predictions = model(x_train[:1]).numpy()
predictions

array([[ 0.68923736, -0.60707337,  0.2112748 ,  0.8276582 , -0.5915252 ,
         0.3560343 ,  0.00231707, -0.1431787 ,  0.00589705, -0.9676335 ]],
      dtype=float32)

__Observation:__
- An array of predicted values is displayed.

## Step 5: Apply softmax activation to the predictions

- Apply softmax activation to the predictions and print the output in terms of probabilities.


In [ ]:
tf.nn.softmax(predictions).numpy()

array([[0.17635825, 0.04824088, 0.10935012, 0.20254017, 0.04899679,
        0.12638266, 0.08872992, 0.07671531, 0.08904814, 0.03363767]],
      dtype=float32)

__Observation:__
- The predicted values in terms of probabilities are displayed as an array.

## Step 6: Define the loss function

- Create a loss function for the model.
- Define the SparseCategoricalCrossentropy loss function.


In [ ]:
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)

In [ ]:
loss_fn(y_train[:1], predictions).numpy()

3.4679463

__Observation:__
- Here, the value is 3.46.

## Step 7: Compile the model

- Compile the model with the Adam optimizer, the loss function, and the accuracy metric.

In [ ]:
model.compile(optimizer='adam',
              loss=loss_fn,
              metrics=['accuracy'])

__Observation:__
- Here, we have used an optimizer as Adam, loss as a loss function, and metrics as accuracy.

## Step 8: Train the model

- Fit the model using the training data for 5 epochs.


In [ ]:
model.fit(x_train, y_train, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 7s 3ms/step - loss: 0.3061 - accuracy: 0.9105
Epoch 2/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.1482 - accuracy: 0.9561
Epoch 3/5
1875/1875 [==============================] - 7s 4ms/step - loss: 0.1101 - accuracy: 0.9666
Epoch 4/5
1875/1875 [==============================] - 10s 5ms/step - loss: 0.0894 - accuracy: 0.9719
Epoch 5/5
1875/1875 [==============================] - 7s 4ms/step - loss: 0.0753 - accuracy: 0.9760


__Observations:__
- The model fits the data.
- Here, we can see that the accuracy is increased by different epochs.

## Step 9: Evaluate the model

- Evaluate the model using the testing data and print the results.


In [ ]:
model.evaluate(x_test,  y_test, verbose=2)

313/313 - 1s - loss: 0.0700 - accuracy: 0.9794 - 775ms/epoch - 2ms/step


[0.07002254575490952, 0.9793999791145325]

__Observation:__

- The accuracy score is 0.979, and the loss is 0.07.

## Step 10: Create a probability model

- Create a probability model by adding the softmax layer to the existing model.



In [ ]:
probability_model = tf.keras.Sequential([
  model,
  tf.keras.layers.Softmax()
])

- Get the predictions using the test data

In [ ]:
probability_model(x_test[:3])

<tf.Tensor: shape=(3, 10), dtype=float32, numpy=
array([[5.5605318e-07, 3.9992972e-08, 1.3901887e-05, 8.1980514e-05,
        1.5747115e-11, 3.8003286e-08, 1.5906798e-12, 9.9989986e-01,
        3.2787732e-06, 2.6140896e-07],
       [1.3507287e-07, 9.2139504e-05, 9.9988079e-01, 2.2869028e-05,
        4.3065286e-14, 2.1160740e-07, 1.5853039e-06, 2.4381213e-13,
        2.2199554e-06, 6.7780311e-17],
       [4.5285432e-07, 9.9852979e-01, 1.8147183e-04, 2.7745891e-05,
        4.6289635e-05, 7.6884826e-06, 5.9110812e-06, 1.1113385e-03,
        8.7311855e-05, 1.9112522e-06]], dtype=float32)>

__Observation:__
- Here, we have the probabilities for the first three observations and different classes.